src/main/netplay.c

```diff
@@ -0,0 +1,339 @@
+    if (check_valid(cin_compat, cin_compat->count))
+        return 1;
+
+    if (l_udpChannel == -1)
+        return 0;
+
+#if SDL_VERSION_ATLEAST(2,0,0)
+    SDL_Thread* thread = SDL_CreateThread(netplay_require_response, "Netplay key request", cin_compat);
+#else
+    SDL_Thread* thread = SDL_CreateThread(netplay_require_response, cin_compat);
+#endif
+    while (!check_valid(cin_compat, cin_compat->count) && l_udpChannel != -1)
+        netplay_process(cin_compat);
+    int success;
+    SDL_WaitThread(thread, &success);
+    return success;
+}
+
+static void netplay_delete_event(struct netplay_event* current, struct controller_input_compat* cin_compat)
+{
+    struct netplay_event* find = cin_compat->event_first;
+    while (find != NULL)
+    {
+        if (find->next == current)
+        {
+            find->next = current->next;
+            break;
+        }
+        find = find->next;
+    }
+    if (current == cin_compat->event_first)
+        cin_compat->event_first = cin_compat->event_first->next;
+    free(current);
+}
+
+BUTTONS netplay_get_input(struct controller_input_compat* cin_compat)
+{
+    BUTTONS keys;
+    netplay_request_input(cin_compat);
+    netplay_process(cin_compat);
+
+    if (buffer_size(cin_compat) > 5)
+    {
+        l_canFF = 1;
+        main_core_state_set(M64CORE_SPEED_LIMITER, 0);
+    }
+    else
+    {
+        main_core_state_set(M64CORE_SPEED_LIMITER, 1);
+        l_canFF = 0;
+    }
+
+    if (netplay_ensure_valid(cin_compat))
+    {
+        struct netplay_event* current = cin_compat->event_first;
+        while (current->count != cin_compat->count)
+            current = current->next;
+        keys = current->buttons;
+        Controls[cin_compat->control_id].Plugin = current->plugin;
+        Controls[cin_compat->control_id].Present = (current->plugin > 0) ? 1 : 0;
+        netplay_delete_event(current, cin_compat);
+        ++cin_compat->count;
+    }
+    else
+    {
+        DebugMessage(M64MSG_ERROR, "Netplay: lost connection to server");
+        main_core_state_set(M64CORE_EMU_STATE, M64EMU_STOPPED);
+        keys.Value = 0;
+    }
+
+    return keys;
+}
+
+void netplay_send_input(struct controller_input_compat* cin_compat, BUTTONS keys, uint8_t Plugin)
+{
+    UDPpacket *packet = SDLNet_AllocPacket(11);
+    packet->data[0] = 0; //Packet type, 0 = send key info
+    packet->data[1] = cin_compat->control_id;
+    SDLNet_Write32(cin_compat->count, &packet->data[2]);
+    SDLNet_Write32(keys.Value, &packet->data[6]);
+    packet->data[10] = Plugin;
+    packet->len = 11;
+    SDLNet_UDP_Send(l_udpSocket, l_udpChannel, packet);
+    SDLNet_FreePacket(packet);
+}
+
+int netplay_register_player(uint8_t player)
+{
+    UDPpacket *requestPacket = SDLNet_AllocPacket(6);
+    UDPpacket *responsePacket = SDLNet_AllocPacket(3);
+
+    requestPacket->data[0] = 4; //Packet type, 4 = player registration
+    requestPacket->data[1] = player;
+    SDLNet_Write32(random_number(), &requestPacket->data[2]);
```

Comment:
I'm not sure to get the idea here. If this number is supposed to match between clients, why not simply use an incremented value? Using rand number could fail on client using various architectures. Am I wrong?

Reply:
Before the game starts, each client needs to register 1 (or more) players to play as. If they don't register any players, they will be a spectator when the game starts. The random number is just a client ID/registration token.

The client sends a message to the server-> "Client (random number) would like to be player 3". Because it is all in UDP, I have to retransmit that message every now and then, using the same random number/client ID. The server will respond with either "yes" or "no" if the client can become that player. Hopefully that makes sense.

I just picked a random number because it was the simplest to implement. Other methods would be to use a MAC address, or some kind of computer ID, but I didn't see any clean, cross platform way to do those things.

Reply:
I just get it. I was thinking this was called all the time. All you say make sense, my bad.